### PR TITLE
Issue #3194775 by Ressinel: Search does not show all results on the same page.

### PR DESCRIPTION
**Pull request by Ressinel**
## Problem
In the resource library a user only sees content that is relevant to their particular permission level which is great.

However, the content does not condense or display well for a user type that cannot see all content (community/logged in user and public user).

As a site manager I can see pretty much all content and it displays nicely with 3 rows of 4 per page.

However, if I see resource library as a logged in user or public user I only see a few resources per page.

Please see screenshots of logged in public user and site manager for reference.

## Solution
Create Search API processor which will be add content access checks for custom view permissions for entities based on field values in the entity_access field. Enable this processor for Search content and Search all indexes.

## Issue tracker
- https://www.drupal.org/project/social/issues/3194775
- https://getopensocial.atlassian.net/browse/YANG-3378

## How to test
- [x] See resource library as a logged in user or public user.

## Screenshots
Before:

for SM:
Screen Shot 2020-07-10 at 12 53 43 PM

for LU and public user:
Screen Shot 2020-07-10 at 12 53 32 PM

After:

for LU and public user:
Screen Shot 2020-07-10 at 12 53 43 PM

## Release notes
Available Search API processor which will be add content access checks for custom view permissions for entities based on field values in the entity_access field.

## Change Record
N/A

## Translations
N/A

### CODE REVIEW

**Review by ribel — APPROVED** [review]
@Ressinel I have tested Resource Library with this new processor and all works as expected now.

@ronaldtebrake, @navneet0693 Initial issue was related for the Resource Library index, but since it is related with `entity_access_by_field` module we decided to add new processor here. 
Also, it might be helpful to add it to Search All and Search Content to fix the similar problem, but I didn't test it and not sure if it needs to be added within this task or follow up. What do you think?

**Comment by ronaldtebrake** [comment]
Hi @ribel, thank for the update.

I think in this case it makes sense to also add it to the other indexes, and since it's access related, I really want this to have test coverage.

Thanks!

**Comment by ribel** [comment]
Hi, @ronaldtebrake, 
As you proposed we fixed it for all search indexes and add test coverage.
Could you check once more if we can merge it now? Also is 9.9 milestone still good here?
And we still have issues with Travis and no ability to re-run jobs correctly :(

**Comment by ronaldtebrake** [comment]
Unfortunately can't help you with travis yet, opened a support ticket there.
However if you look at the tests
https://travis-ci.com/github/goalgorilla/open_social/jobs/482300126#L2190

Indicates there might be something wrong? Have you tried running these locally?

**Comment by Ressinel** [comment]
Hi, @ronaldtebrake 
I will try rerun test local.

**Comment by ribel** [comment]
@ronaldtebrake @Ressinel I have run tests locally and they passed. 
image

**Comment by Ressinel** [comment]
Hi @ribel @ronaldtebrake 

Yes, the same situation.

**Comment by ronaldtebrake** [comment]
Update milestone to 10.1 based on the fact we are adding new processor so it's a bit much for a minor release in 9.x.

**Comment by ronaldtebrake** [comment]
Hi @Ressinel @ribel 

Thanks guys, although I'm a bit confused as to how this works or if this is the correct solution.
After checking it out locally I have been checking search api (and solr for that matter) to see what was going on with this new additional processor.

Looking in the database I can see the following after re-indexing. (I was checking 2 entities created, one in an open group one in a closed)

Screenshot 2021-02-26 at 12 06 05

So with this PR it adds the new entity access by fields as a separate index.

However if I check these same items in node_grants

Screenshot 2021-02-26 at 12 05 53

These are also already there, so is this new processor even necessary? 🤔 

Now we get duplicate amount of grants in our indexes which increases our size quite drastically.

Can you maybe ellaborate why this is necessary?

File: modules/custom/entity_access_by_field/src/Plugin/search_api/processor/EntityAccessByField.php [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+    }
+
+    // Gather the affected datasources, grouped by entity type, as well as the
+    // unaffected ones.
+    $affected_datasources = [];
+    $unaffected_datasources = [];
+    foreach ($this->index->getDatasources() as $datasource_id => $datasource) {
+      $entity_type = $datasource->getEntityTypeId();
+      if (in_array($entity_type, ['node'])) {
+        $affected_datasources[$entity_type][] = $datasource_id;
```
Comment (ronaldtebrake) [inline modules/custom/entity_access_by_field/src/Plugin/search_api/processor/EntityAccessByField.php:298]: This doesn't seem to be used anywhere in this context.

**Comment by Ressinel** [comment]
Hi @ronaldtebrake 

That processor firstly was created for Resource library. Then I noticed that we have the same issue for other indexes.
I will try to implement your suggestions you provided.

Thanks!

**Comment by sjoerdvandervis** [comment]
Might be wise to give this a bit of priority to merge it, as it fixes an issue with no resources shown in the RL. This probably has to be merged together with the pull request for the resource library [here](https://bitbucket.org/goalgorilla/social_resource_library/pull-requests/33) . Of course, the remarks from @ronaldtebrake are still valid and need to be taken care of.

**Comment by Ressinel** [comment]
Hi @ronaldtebrake 
I wanted to report that I have started to check this PR. I Will keep you updated.

**Comment by Ressinel** [comment]
@ronaldtebrake Did you check the same scenario with SOLR Server?

**Comment by ronaldtebrake** [comment]
I think it had the same maybe @navneet0693 can confirm, wasn't 100% sure on that

**Comment by Ressinel** [comment]
Hi @navneet0693 

Check the comment above, please.

Thanks!

**Comment by navneet0693** [comment]
> Hi @navneet0693
> 
> Check the comment above, please.
> 
> Thanks!

Yes, we did check on SOLR.

**Comment by Ressinel** [comment]
@ronaldtebrake 

I have updated the new processor.
See how there will be an opportunity, please.

Thanks!

**Comment by ronaldtebrake** [comment]
Thanks for the work guys, I have asked @navneet0693 to verify this on one of our environments where we encountered this before 👍

**Comment by Ressinel** [comment]
Hi, @ronaldtebrake @navneet0693 

Can you give feedback for the PR last changes, please?

Thanks!

**Comment by ronaldtebrake** [comment]
Hi @Ressinel 

I've just tested this, it seems to be almost working, not entirely sure if it's connected to your PR or not maybe you can enlighten us.

Let me paint the scenario:

I've added a topic, set to group members, in a Closed Group.
I've added a topic, set it to group members, in a Flexible group with content visibility options community & group members

In the below screenshot you see a LU on the search content, only sorting on A - Z (so i can easily compare results with a different user) I expect there to be 10 results per page only 9 are shown (it seems to be due to the topic in the flexible group, the topic in the closed group is correctly removed from the results and pagination total otherwise we would have seen only 8 results if that didn't work.. so maybe that processor also needs to correctly check flexible group options?)

Screenshot 2021-03-19 at 11 46 00


For a SM+ you see the below results where you see both topics with topic title "Flexible group topic - group members visibility" and "Closed group topic - Members visibility" which are the topics in the flexible group and closed group.

Screenshot 2021-03-19 at 11 51 43

So it seems we're only missing the flexible group visibility part.

**Comment by Ressinel** [comment]
Thanks for your response and investigation @ronaldtebrake 

Has it been tested with an enabled EntityAccessByField processor?

**Comment by ronaldtebrake** [comment]
Hi @Ressinel 
A-ha, thanks for pointing that out :D I missed that, it wasn't enabled indeed although I did run the update hooks which passed nicely and I expected those to enable it for me and do a re-index.

Screenshot 2021-03-19 at 12 30 55
Screenshot 2021-03-19 at 12 30 47


Is that something you can still take a look at, after manually enabling and re-indexing that definitely worked! 

Thanks for that 👍

**Comment by Ressinel** [comment]
@ronaldtebrake

If I right understand processor looks good, but I need to make sure it turns on when I run the update hooks, right?

**Comment by ronaldtebrake** [comment]
@Ressinel :D you are correct!

**Comment by Ressinel** [comment]
Hi, @ronaldtebrake

I tested to run the update hooks and the processor turn on for me after it.
Can you, maybe, try to run the update hooks from this PR on empty env? I mean where updates on this PR have not been launched yet.

Thanks!

**Comment by ronaldtebrake** [comment]
@Ressinel - thanks - Seems to work now, not sure why it didnt before 👍

**Comment by Ressinel** [comment]
Hi, @ronaldtebrake 
Is this PR ready to merge?

**Comment by ronaldtebrake** [comment]
If we can just merge 10.1.x and see if all tests will turn green, I'm happy @Ressinel 🚀

**Pull request merged**